File: HW02_B0829002/partition0/app_checkChain.go
package main

import (
	"crypto/sha256"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
	"os/exec"
)

// checkBlock returns the current highest block number by checking the filesystem.
func checkBlock() int {
	i := 1
	for {
		if _, err := os.Stat(strconv.Itoa(i) + ".txt"); os.IsNotExist(err) {
			break
		}
		i++
	}
	return i
}

func main() {
	// Checking the number of arguments
	if len(os.Args) != 2 || os.Args[1] == "" {
		fmt.Println("Usage: go run app_checkChain.go or go run app_checkChain.go username")
		os.Exit(1)
	}

	currentBlock := checkBlock()
	var failedBlocks []int

	for i := 1; i < currentBlock-1; i++ {
		// Reading the entire content of the current block file
		content, err := ioutil.ReadFile(strconv.Itoa(i) + ".txt")
		if err != nil {
			fmt.Println("Error reading file:", err)
			continue
		}
		lineAll := string(content)

		// Calculating the hash of the current block
		hashObject := sha256.Sum256([]byte(lineAll))
		hashString := fmt.Sprintf("%x", hashObject)

		// Reading the next block to compare the stored hash
		nextBlockContent, err := ioutil.ReadFile(strconv.Itoa(i+1) + ".txt")
		if err != nil {
			fmt.Println("Error reading next file:", err)
			continue
		}
		lines := strings.Split(string(nextBlockContent), "\n")
		firstLine := lines[0]
		firstLineParts := strings.Split(firstLine, " ")
		if len(firstLineParts) < 5 {
			fmt.Println("Invalid format in next block's first line.")
			continue
		}
		storedHash := firstLineParts[4]
		// Removing potential trailing characters from the stored hash
		storedHash = strings.TrimSpace(storedHash)

		if hashString != storedHash {
			fmt.Printf("The blockchain has been tampered with at block %d\n", i)
			failedBlocks = append(failedBlocks, i)
		}
	}

	if len(failedBlocks) == 0 {
		fmt.Println("The blockchain is safe")
		if len(os.Args) == 2 {
			cmd := exec.Command("./app_transaction", "angel", os.Args[1], "5")
			cmd.Stdout = os.Stdout
			cmd.Stderr = os.Stderr
			cmd.Run()
		}
	} else {
		fmt.Println("The following blocks have been tampered with:", failedBlocks)
	}
}

File: HW01/app_checkChain.py
from library import checkBlock
import hashlib
import sys, subprocess

if len(sys.argv) != 2 and len(sys.argv) != 1:
    print("Usage: python3 app_checkChain.py [username] or python3 app_checkChain.py")
    sys.exit()

currentBlock = checkBlock()
lineAll = ""
hash_object = ""
failed_block = []
for i in range(1, currentBlock-1):
    lineAll = ""
    print("Block: {}.txt".format(i))
    f = open("{}.txt".format(i), "r")
    lines = f.readlines()
    for line in lines:
        lineAll += line
    hash_object = hashlib.sha256(lineAll.encode())
    f.close()
    # oepn the next block
    f = open("{}.txt".format(i+1), "r")
    # get the first line and compare the hash
    lines = f.readlines()
    firstLine = lines[0]
    firstLine = firstLine.split(" ")[4]
    firstLine = firstLine[:-1]
    # compare the hash
    print(hash_object.hexdigest())
    print(firstLine)
    if hash_object.hexdigest() != firstLine:
        print("The blockchain has been tampered with {}".format(i))
        failed_block.append(i)
if len(failed_block) == 0:
    print("The blockchain is safe")
    if len(sys.argv) == 2:
        user = sys.argv[1]
        # execute python3 app_checktransaction.py angel username 5
        subprocess.call(["python3", "app_transaction.py", "angel", user, "5"])
        
else:
    print("The following blocks have been tampered with: {}".format(failed_block))